### ISSUE TITLE
CacheCow.Client seems not to be used (.NET core)

### ISSUE THREAD

**Issue by jagbarcelo** [aliostad/CacheCow#238]
The problem we are seeing is that the server seems to be returning right responses (with Cache-Control, ETag,... even x-cachecow-server headers) but the client never makes a conditional get (those with If-None-Match, or If-Modified-Since headers). Because of this, the server is always replying with full new responses (200 Ok) and never a (304 Not modified).

The only special thing we have is that we want the server to reply to authenticated requests as cacheable too. In order to achieve that we have created `MyCacheabilityValidator` based on your original `DefaultCacheabilityValidator` and commented out the lines about authenticated requests:

```
public bool IsCacheable(HttpRequest request)
{
    // none-GET (HEAD is ignored here)
    if (!request.Method.Equals("GET", StringComparison.InvariantCultureIgnoreCase))
        return false;

    // MY CHANGES: We want to be able to cache auth'ed requests too!!!
    //// auth
    //if (request.Headers.Any(x => x.Key.Equals(HttpHeaderNames.Authorization, StringComparison.InvariantCultureIgnoreCase)))
    //    return false;

    // pragma no-cache
    if (request.Headers.Any(x => x.Key.Equals(HttpHeaderNames.Pragma, StringComparison.InvariantCultureIgnoreCase)) &&
        request.Headers[HttpHeaderNames.Pragma].Any(x => x.Contains("no-cache")))
        return false;

    if (request.Headers.Any(x => x.Key.Equals(HttpHeaderNames.CacheControl, StringComparison.InvariantCultureIgnoreCase)) &&
        request.Headers[HttpHeaderNames.CacheControl].Any(x => x.Contains("no-store")))
        return false;

    return true;
}
```

Why is the client never requesting a conditional get? Might it be because we need to do Dependency Injection in the client side too in order to override the default behavior of not-caching authenticated queries (despite whatever the server side says)? Previously (when we were using WebApi) we only needed this fix to be applied at the server side.

In the client side we have some middleware as:

```
var pipeline = new CustomDelegatingHandler(LicenseFile)
{
    InnerHandler = new CacheCow.Client.CachingHandler()
    {
        InnerHandler = new HttpClientHandler() { AutomaticDecompression = DecompressionMethods.All }
    }
};
MyHttpClient = new HttpClient(pipeline);
```

The first CustomDelegatingHandler is in charge of inserting the Authorization headers in the requests, based on the information provided in the LicenseFile. The second one is your CacheCow.Client and the third one is for using compression (GZip, Deflate or Brotli).

Previously, in our WebApi version (which worked perfectly), we had this in the client side (it seems equivalent to me, although a bit less visual):

```
var compressionHandler = new HttpClientHandler();
if (compressionHandler.SupportsAutomaticDecompression) compressionHandler.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
var cachingHandler = new CacheCow.Client.CachingHandler() { InnerHandler = compressionHandler };
MyHttpClient = HttpClientFactory.Create(cachingHandler, new DelegatingHandler[] { new CustomDelegatingHandler(LicenseFile) });
```

A sample request / response of the new version (.NET core) which never makes conditional requests.

```
GET https://localhost:44370/api/v1.0/user/1ca7c112-dde2-e571-e081-90155d0b0711/project/paging?useDateCreated=True&year=2019&month=9&pageSize=25 HTTP/1.1
Host: localhost:44370
Accept: application/json
Accept-Language: es-ES
Cache-Control: max-age=86400, private
User-Agent: MYAPP-DEBUG/2.1.1.0
Authorization: amx 23e07635-783f-4dc2-a6a4-1cbc4036fb21:86F1-ACCD-8331-7E90-A9F8-CC3F-2A8C-09A2:KMaQVNXi8pHzV8UnRQ8bKOnoTrFf22ZfwDTysycB4b0=:c3e4ae5c29604f7da0ddfe73bca5e232:1571384223
Accept-Encoding: gzip, deflate, br


HTTP/1.1 200 OK
Cache-Control: public, must-revalidate, max-age=86400
Transfer-Encoding: chunked
Content-Type: application/json; charset=utf-8
Content-Encoding: br
ETag: "a4ayc/80/OGda4BO/1o/V0etpOqiLx1JwB5S3beHW0s="
Vary: Accept,Accept-Encoding
Server: Microsoft-IIS/10.0
x-cachecow-server: validation-applied=False;validation-matched=False;short-circuited=False;query-made=False
X-Powered-By: ASP.NET
Date: Fri, 18 Oct 2019 07:37:02 GMT
[Body Removed]
```

Any ideas about why the client seems not to be using CacheCow.Client at all? Thanks again for your help and time.

**Comment by jagbarcelo** [comment]
A little more feedback. I have just added the following lines for debugging just the client (against a well known public server):


```
var compressionHandler = new HttpClientHandler();
if (compressionHandler.SupportsAutomaticDecompression) compressionHandler.AutomaticDecompression = DecompressionMethods.All;

var pipeline = new CustomDelegatingHandler(LicenseFile)
{
    InnerHandler = new CacheCow.Client.CachingHandler()
    {
        InnerHandler = compressionHandler
    }
};
MyHttpClient = new HttpClient(pipeline);

// USED TO TEST/CHECK IF CACHECOW.CLIENT WORKS.
const string CacheableResource = "https://code.jquery.com/jquery-3.3.1.slim.min.js";
var response = MyHttpClient.GetAsync(CacheableResource).ConfigureAwait(false).GetAwaiter().GetResult();
var responseFromCache = MyHttpClient.GetAsync(CacheableResource).ConfigureAwait(false).GetAwaiter().GetResult();
System.Diagnostics.Debug.Print(response.Headers.ToString());
System.Diagnostics.Debug.Print(responseFromCache.Headers.ToString());

```
The second response is made right against the same resource so it should be in the client cache and the client should have made a conditional get. Here is the output of those Debug.Print's:


```
Date: Fri, 18 Oct 2019 09:42:32 GMT
Connection: Keep-Alive
Accept-Ranges: bytes
Server: nginx
ETag: W/"5a637bd4-1111d"
Cache-Control: public, max-age=315360000
Access-Control-Allow-Origin: *
Vary: Accept-Encoding
X-HW: 1571391752.dop042.ma1.t,1571391752.cds030.ma1.shn,1571391752.dop042.ma1.t,1571391752.cds017.ma1.c
x-cachecow-client: 2.4.4.0;did-not-exist=true

Date: Fri, 18 Oct 2019 09:42:36 GMT
Connection: Keep-Alive
Accept-Ranges: bytes
Server: nginx
ETag: W/"5a637bd4-1111d"
Cache-Control: public, max-age=315360000
Access-Control-Allow-Origin: *
Vary: Accept-Encoding
X-HW: 1571391752.dop042.ma1.t,1571391752.cds030.ma1.shn,1571391752.dop042.ma1.t,1571391756.cds017.ma1.c
x-cachecow-client: 2.4.4.0;did-not-exist=true

```

As you can see, it has the x-cachecow-client headers, so it seems the cache is being used. However no conditional get is made for the second request. For your tests, you can safely skip/ignore the usage of our CustomDelegatingHandler middleware, which does nothing when the request is not done against our own server (it only adds the Authorization header when necessary).

Regards.

**Comment by aliostad** [comment]
When I set the compression to None it works. It could be a bug when compression is on. 

Thanks I will look into it.

```
Date: Fri, 18 Oct 2019 12:00:13 GMT
Connection: Keep-Alive
Accept-Ranges: bytes
Server: nginx
ETag: W/"5a637bd4-1111d"
Cache-Control: public, max-age=315360000
Access-Control-Allow-Origin: *
Vary: Accept-Encoding
X-HW: 1571400013.dop017.lo4.t,1571400013.cds061.lo4.shn,1571400013.cds061.lo4.c
x-cachecow-client: 2.5.1.0;did-not-exist=true

Date: Fri, 18 Oct 2019 12:00:13 GMT
Connection: Keep-Alive
Accept-Ranges: bytes
Server: nginx
ETag: W/"5a637bd4-1111d"
Cache-Control: public, max-age=315360000
Access-Control-Allow-Origin: *
Vary: Accept-Encoding
X-HW: 1571400013.dop017.lo4.t,1571400013.cds061.lo4.shn,1571400013.cds061.lo4.c
x-cachecow-client: 2.5.1.0;did-not-exist=false;retrieved-from-cache=true

```

**Comment by aliostad** [comment]
OK it works under .NET 4.5.2 and .NET Core 2.0 but not 2.1 or 3.0. (Please confirm)

They have messed it up, I have a couple of open issues with .NET team not yet sorted this is related to it.

**Comment by jagbarcelo** [comment]
Thanks for your support. I suppose we'll need to stuck either with caching or with compression (but not both) while this is solved.

**Comment by aliostad** [comment]
What about dropping to .net core 2.0? Probably not an option?

**Comment by aliostad** [comment]
OK it was a weird issue which happens due to the changes to in .NET Core 2.1+. I have made a change and published with version 2.5.2

It is not easy to explain the case but I will try:

Essentially the server sends `Vary` header which value "Accept-Encoding". This means that the body sent back by the server changes based on the value of "Accept-Encoding" header. CacheCow uses the value of Accept-Encoding header as part of its cache key, since content changes by that.

Now, you do not add that header but as of .NET Core 2.1 and later, setting `compressionHandler.AutomaticDecompression` will result in adding Accept-Encoding header. Problem is CacheCow does not see it since it is added by the client handler which is a deeper handler. This is the behaviour change (and not a bug) in 2.1 and 3.0 which creates this problem.

So the solution was that I keep the _original_ request headers and use those for creating the cache key so it is consistent.

Generally in this case, I suggest that you set the DefaultVaryHeaders in your ClientHandler to both "Accept" and "Accept-Encoding" especially now that you are dealing with compression. One word of caution: for reasons that it is even harder to explain, caching a resource might kick in after the 3rd request (not the normal 2nd) if server sends `Vary` headers that are not in your default vary headers - which is not the end of the world but might trip you up while debugging. 

You can have a look at the test I have [created](https://github.com/aliostad/CacheCow/blob/master/test/CacheCow.Client.Tests/CachingHandlerTests.cs#L521) for this issue - which now passes.
